lang/py/src/avro/schema.py

```diff
@@ -405,6 +419,102 @@ def __eq__(self, that):
     return self.props == that.props
 
 #
+# Decimal Type
+#
+class DecimalSchema(LogicalSchema):
+  def __init__(self, decimal_type, precision, scale):
+    if decimal_type != 'fixed' and decimal_type != 'bytes':
+      raise SchemaParseException('Logical type DECIMAL must be backed by fixed or bytes.')
+
+    LogicalSchema.__init__(self, decimal_type, 'decimal')
+    self.set_prop('precision', precision)
+    self.set_prop('scale', scale)
+
+  def validate(self, validation_props=None):
+    if not isinstance(validation_props.get('precision'), int):
+      raise SchemaParseException('Precision (Integer) is required for logical type DECIMAL.')
+
+    precision = validation_props.get('precision')
+    if precision <= 0:
+      raise SchemaParseException("Invalid DECIMAL precision %s. Must be positive."
+                                 % precision)
+    elif precision > self._max_precision(validation_props):
+      raise SchemaParseException("Fixed %s cannot store precision digits. Max is %s"
+                                 %(validation_props.get('size'),
+                                   self._max_precision(validation_props)))
+
+    if not isinstance(validation_props.get('scale'), int):
+      raise SchemaParseException('Scale must be of type Integer.')
+
+    scale = validation_props.get('scale')
+    if scale < 0:
+      raise SchemaParseException("Invalid DECIMAL scale %s. Must be positive"
+                                 % scale)
+    elif scale > precision:
+      raise SchemaParseException("Invalid DECIMAL scale %s. Cannot be greater than precision %s"
+                                 %(scale, precision))
+
+  def _max_precision(self, validation_props):
+    decimal_type = validation_props.get('decimal_type')
+    if decimal_type == 'bytes':
+      # Considering the max 32 bit integer value
+      return (1 << 31) - 1
+    elif decimal_type == 'fixed':
+      return round(floor(log10(pow(2, 8 * validation_props.get('size') - 1) - 1)))
+    else:
+      # Since DECIMAL is not valid for any other type
+      return 0
```

Comment:
This `max_precision` really should be provided by the subclasses (`BytesDecimalSchema` and `FixedDecimalSchema`)